x-pack/plugins/fleet/public/applications/fleet/sections/agents/agent_details_page/components/agent_logs/index.tsx

```diff
@@ -0,0 +1,234 @@
+
+  // Util to convert date expressions (returned by datepicker) to timestamps (used by LogStream)
+  const getDateRangeTimestamps = useCallback(
+    (timeRange: TimeRange) => {
+      const { min, max } = data.query.timefilter.timefilter.calculateBounds(timeRange);
+      return min && max
+        ? {
+            startTimestamp: min.valueOf(),
+            endTimestamp: max.valueOf(),
+          }
+        : undefined;
+    },
+    [data.query.timefilter.timefilter]
+  );
+
+  // Initial time range filter
+  const [dateRange, setDateRange] = useState<{
+    startExpression: string;
+    endExpression: string;
+    startTimestamp: number;
+    endTimestamp: number;
+  }>({
+    startExpression: DEFAULT_DATE_RANGE.start,
+    endExpression: DEFAULT_DATE_RANGE.end,
+    ...getDateRangeTimestamps({ from: DEFAULT_DATE_RANGE.start, to: DEFAULT_DATE_RANGE.end })!,
+  });
+
+  const tryUpdateDateRange = useCallback(
+    (timeRange: TimeRange) => {
+      const timestamps = getDateRangeTimestamps(timeRange);
+      if (timestamps) {
+        setDateRange({
+          startExpression: timeRange.from,
+          endExpression: timeRange.to,
+          ...timestamps,
+        });
+      }
+    },
+    [getDateRangeTimestamps]
+  );
+
+  // Filters
+  const [selectedLogLevels, setSelectedLogLevels] = useState<string[]>([]);
+  const [selectedDatasets, setSelectedDatasets] = useState<string[]>([AGENT_DATASET]);
+
+  // User query state
+  const [query, setQuery] = useState<string>('');
+  const [draftQuery, setDraftQuery] = useState<string>('');
+  const [isDraftQueryValid, setIsDraftQueryValid] = useState<boolean>(true);
+  const onUpdateDraftQuery = useCallback((newDraftQuery: string, runQuery?: boolean) => {
+    setDraftQuery(newDraftQuery);
+    try {
+      esKuery.fromKueryExpression(newDraftQuery);
+      setIsDraftQueryValid(true);
+      if (runQuery) {
+        setQuery(newDraftQuery);
+      }
+    } catch (err) {
+      setIsDraftQueryValid(false);
+    }
+  }, []);
+
+  // Base queries: agent id, dataset, log level
+  const AGENT_QUERY = useMemo(() => {
+    const agentQuery = `${AGENT_ID_FIELD.name}:${agent.id}`;
+    const datasetQuery = selectedDatasets.length
+      ? selectedDatasets.map((dataset) => `${DATASET_FIELD.name}:${dataset}`).join(' or ')
+      : `${DATASET_FIELD.name}:${AGENT_DATASET} or ${DATASET_FIELD.name}:${AGENT_DATASET}.*`;
+    return `${agentQuery} and (${datasetQuery})`;
+  }, [agent.id, selectedDatasets]);
+  const LOG_LEVEL_QUERY = useMemo(
+    () => selectedLogLevels.map((level) => `${LOG_LEVEL_FIELD.name}:${level}`).join(' or '),
+    [selectedLogLevels]
+  );
+  const BASE_QUERY = useMemo(
+    () => (LOG_LEVEL_QUERY ? `(${AGENT_QUERY}) and (${LOG_LEVEL_QUERY})` : AGENT_QUERY),
+    [AGENT_QUERY, LOG_LEVEL_QUERY]
+  );
+
+  // Final query: base + user queries
+  const LOG_STREAM_QUERY = useMemo(() => (query ? `(${BASE_QUERY}) and (${query})` : BASE_QUERY), [
+    BASE_QUERY,
+    query,
+  ]);
```

Comment:
This part seems easy to break, seems a good candidate for a unit test what do you think? 


also nitpick, I have preference to use camelCase for local variable and only uppercase for top level constant

Reply:
good call, I've moved the query building logic into a separate file with a unit test